lib/ansiblelint/errors.py

```diff
@@ -12,10 +15,14 @@ def __init__(self, message=None, linenumber=0, line=None, filename=None, rule=No
         """Initialize a MatchError instance."""
         super().__init__(message)
 
+        if not (message or rule):
+            self.linenumber = 0
+            raise RuntimeError("Calling MatchError requires either a message or a rule.")
```

Comment:
Technically, it's not "calling" but initializing (or raising).
Also, I'd like to follow Python's own conventions and raise the same type of exception that is `TypeError`:
```python
In [1]: def some_func(arg): ...                                               

In [2]: some_func(1)                                                          

In [3]: some_func()                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-3d9630d20b3d> in <module>
----> 1 some_func()

TypeError: some_func() missing 1 required positional argument: 'arg'


In [4]: def some_func_w_kwarg_only(arg, *, kwarg): ...                        

In [6]: some_func_w_kwarg_only(1, kwarg=2)                                    

In [7]: some_func_w_kwarg_only(1)                                             
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-861031c984e2> in <module>
----> 1 some_func_w_kwarg_only(1)

TypeError: some_func_w_kwarg_only() missing 1 required keyword-only argument: 'kwarg'

In [8]: some_func_w_kwarg_only(kwarg=2)                                       
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-bca0418a0e53> in <module>
----> 1 some_func_w_kwarg_only(kwarg=2)

TypeError: some_func_w_kwarg_only() missing 1 required positional argument: 'arg'

In [9]: class Demo: 
   ...:     def __init__(self, arg, *, kwarg): ... 
   ...:                                                                       

In [10]: Demo(1, kwarg=2)                                                     
Out[10]: <__main__.Demo at 0x7fe29886c2e8>

In [11]: Demo(kwarg=2)                                                        
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-5b16a6841949> in <module>
----> 1 Demo(kwarg=2)

TypeError: __init__() missing 1 required positional argument: 'arg'

In [12]: Demo(1)                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-69c1ae1e2087> in <module>
----> 1 Demo(1)

TypeError: __init__() missing 1 required keyword-only argument: 'kwarg'
```

And so the better exception would be
```suggestion
            raise TypeError(
                f'{self.__class__.__name__}() missing a '
                "required argument: one of 'message' or 'rule'",
            )
```